BugFix: Basic way to save ragged arrays with ndim>1

CSSFrancis opened the pull request:
### This is a pretty basic way to fix the saving of signals with a ragged dim greater than 2

The motivation for this is with saving vector signals after some analysis.  Previously these signals weren't saving properly (see #2904)

**_The general procedure:_**
- Unwrapped ragged arrays with signal dim greater than two
- Save the original shape of the signal dim as a separate dataset 
- Load the unwrapped data and original shape
- Reshape the data using the original shape.

### Progress of the PR
- [x]  Change implemented (can be split into several points),
- [ ] update docstring (if appropriate),
- [ ] update user guide (if appropriate),
- [ ] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)),
- [ ] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [ ] add tests,
- [ ] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
import numpy as np

x = np.empty(5,dtype=object)
for i in range(1,6):
    x[i-1]= np.array([list(range(i)),list(range(i))])

s = hs.signals.BaseSignal(x, ragged=True)

s.save("test.hspy")

s =hs.load("test.hspy")
```

@ericpre  Any suggestions you have here would be greatly appreciated.  Right now it works and is only a rather small change.  That being said it am not extremely happy with how I got it to work.

CSSFrancis committed:
BugFix: Basic way to save ragged arrays with ndim>1

ericpre commented on hyperspy/io_plugins/_hierarchical.py:
```diff
@@ -626,7 +634,17 @@ def overwrite_dataset(cls, group, data, key, signal_axes=None,
 
         _logger.info(f"Chunks used for saving: {chunks}")
-        cls._store_data(data, dset, group, key, chunks)
+        if data.dtype == np.dtype('O'):
```
To enforce declaring signal as being ragged, I think we should use this approach only when the signal is ragged:
```suggestion
        if self.signal.ragged:
```

CSSFrancis replied:
I don't know if that is the best behavior right now.  Currently the ragged attribute isn't firmly typed i.e. you can have a signal which has an object dtype but isn't ragged.  It probably shouldn't be the case going forward but for now I think that would cause some edge case signal to not save.

ericpre replied:
Yes, this is the motivation to enforce the use of `ragged` signal when necessary and special case `ragged` processing only when the signal is explicitly set as ragged, otherwise this is speculative.

If I understand/remember correctly, the issue is with saving ragged array, not `object` type array? If so, I think that using `self.signal.ragged` is more appropriate.

ericpre commented on hyperspy/io_plugins/_hierarchical.py:
```diff
@@ -626,7 +634,17 @@ def overwrite_dataset(cls, group, data, key, signal_axes=None,
+            new_data = np.empty(shape=data.shape, dtype=object)
+            shapes = np.empty(shape=data.shape, dtype=object)
+            for i in np.ndindex(data.shape):
+                new_data[i] = np.ndarray.flatten(data[i])
```
```suggestion
                new_data[i] = data[i].flatten()
```

CSSFrancis replied:
Is there a reason for doing it this way instead?

ericpre replied:
The method may be marginally faster than the function but the main reason is that it is shorter to write and read!

Actually, in this case, it may be better to use `ravel` to return a view and avoid making a copy.

ericpre reviewed (commented):
Thanks @CSSFrancis, this approach looks good to me.

ericpre commented:
I edited the example [above](https://github.com/hyperspy/hyperspy/pull/2906#issue-1173122868) to the recommend way of creating signals of ragged array as per https://hyperspy.readthedocs.io/en/latest/user_guide/signal.html#ragged-array.

CSSFrancis committed:
Testing: Changed test to inline with documentation method for creating ragged signals.

Testing: Added testing for saving 2 dim ragged arrays and fixed broken hyperspy saving.

CSSFrancis committed:
Documentation: Added to changelog.

ericpre committed:
Use ravel instead of flatten to avoid making copy. Remove unused code.
